### PR TITLE
[WIP] enable cuda graphs support for flash attention with dropout

### LINKED ISSUE #99905: Dropout support for SDPA with torch.compile
### Eager Mode
Currently when someone calls https://github.com/pytorch/pytorch/blob/f0e28b1cb9e0bc30877465b9c804a46461e537db/aten/src/ATen/native/native_functions.yaml#L14032 with `dropout_p != 0.0` in eager mode everything works great. FlashAttention is somewhat anomalous in its RNG handiling. The function will "checkpoint" the current state of the RNGEngine when calling the forward pass. It will NOT save the full attention mask for backward and instead save (for cuda) the philox_seed and philox_offset. When not using CudaGraphs these two int64_t can be used to fully capture the state of the PhiloxRNGEngine. The code that does this in flash_mha_fwd can be found https://github.com/pytorch/pytorch/blob/f0e28b1cb9e0bc30877465b9c804a46461e537db/aten/src/ATen/native/transformers/cuda/flash_attn/fmha_api.cpp#L321

In the backward function a temporary PhiloxState is constructed from the seed and offset saved for bacwkard https://github.com/pytorch/pytorch/blob/f0e28b1cb9e0bc30877465b9c804a46461e537db/aten/src/ATen/native/transformers/cuda/flash_attn/fmha_api.cpp#L498

Notice that I added a `TORCH_CHECK(at::cuda::currentStreamCaptureStatus() == at::cuda::CaptureStatus::None` This is because when using cuda_graphs more information that two ints is needed to fully rebuild the state.  Ideally it would be easy to pass around a convenient representation of the PhiloxState from forward to backward functions. I would like to find a solution for enabling FlashAttention with CudaGraphs.

### Torch Compile
There are two issues currently with using sdpa_flash_attention w/ `dropout_p != 0.0` and torch compile.
1.) The CudaGraph issue mentioned above.
2.) At the time of creating this PR: https://github.com/pytorch/pytorch/pull/92917 the two integers `philox offset` and `philox seed` were getting baked into the compiled graph.

For 2) I left a note in the meta_regsitrations.py file here: https://github.com/pytorch/pytorch/blob/39590d06c563d830d02b9f94611ab01f07133c97/torch/_meta_registrations.py#L2350

After the fact Ed and Alban mentioned that https://github.com/pytorch/pytorch/blob/39590d06c563d830d02b9f94611ab01f07133c97/torch/_meta_registrations.py#L2417 Could be replaced with a symint directly and that would likely fix the issue?

This code was added to enforce graph_break https://github.com/pytorch/pytorch/blob/39590d06c563d830d02b9f94611ab01f07133c97/torch/_dynamo/variables/torch.py#L513
Causing graph breaks is shown to have a noticable inmpact on the performance of a compiled module. I expereimented with NanoGPT and produced the following results:

Eager TImes
```
FlashAttention Kernel dropout=0.0: iter 3: loss 10.8671, time 902.01ms
FlashAttention Kernel dropout=0.1: iter 3: loss 10.8671, time 900.97ms
```
Compile Times
```
FlashAttention Kernel dropout=0.0 w/ torch compile: iter 3: loss 10.8674, time 552.06ms
FlashAttention Kernel dropout=0.1 w/ torch compile: iter 3: loss 10.8830, time 610.35ms
```

### Questions
I am planning on  upstreaming the recent changes from Xformers to mem-efficient attention. They have also enabled dropout support using a similar RNG checkpoint mechanism. I would like come up with a solution that better handles dropout with pt 2.0

How should I PhiloxState be passed from forward to backward functions? How can the meta function be updated to not bake in the offsets and seeds?  Can these changes be FC/BC?

cc @jbschlosser @bhosmer @cpuhrsch @erichan1 @ezyang @soumith @msaroufim @wconstab @ngimel @bdhirsh

### PR DESCRIPTION
**Pull request by ngimel**
Fixes #99905

cc @soumith @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

### CODE REVIEW

File: aten/src/ATen/native/transformers/cuda/flash_attn/fmha_api.cpp [hunk 1]
```diff
@@ -318,20 +320,22 @@ mha_fwd(const at::Tensor &q,         // total_q x num_heads x head_size, total_q
...
+    at::Tensor seed_t, offset_t;
+    if (is_dropout) {
         TORCH_CHECK(at::cuda::currentStreamCaptureStatus() == at::cuda::CaptureStatus::None,
         "scaled_dot_product_flash_attention does not support dropout with cuda graph capture mode enabled");
         // See Note [Acquire lock when using random generators]
         std::lock_guard<std::mutex> lock(gen->mutex_);
         // generator_state = at::Tensor::wrap_tensor_impl(gen -> get_state());
         at::PhiloxCudaState philox_state = gen->philox_cuda_state(counter_offset);
         std::tie(seed, offset) = at::cuda::philox::unpack(philox_state);
+        seed_t = at::scalar_tensor(at::Scalar(static_cast<int64_t>(seed)), at::dtype(at::kLong));
```
Comment (drisspg) [inline aten/src/ATen/native/transformers/cuda/flash_attn/fmha_api.cpp:332]: does this need to be device=CUDA if we are in graph capture?

File: aten/src/ATen/native/transformers/cuda/flash_attn/fmha_fprop_kernel_1xN.h [hunk 2]
```diff
@@ -680,6 +680,11 @@ inline __device__ void device_1xN_loop(const Params &params) {
     // the attention matrix. This way, as long as we have the batch, head, and the location of
     // the 16 x 16 block within the attention matrix, we can generate the exact same dropout pattern.
     auto seeds = at::cuda::philox::unpack(params.philox_args);
+    if (params.philox_args.captured_) {
+        *params.seed = std::get<0>(seeds);
+        *params.extragraph_offset = std::get<1>(seeds);
+    }
+
```
Comment (drisspg) [inline aten/src/ATen/native/transformers/cuda/flash_attn/fmha_fprop_kernel_1xN.h:687]: I feel like this information is no propgating forward, like the next line should be:
`Philox ph(std::get<0>(seeds), 0, std::get<1>(seeds) + extragraph_offset + (bidb * params.h + bidh) * 32 + tidx % 32);`

But not sure about the first seed

File: aten/src/ATen/native/nested/cuda/NestedTensorTransformerFunctions.cpp [hunk 3]
```diff
@@ -710,8 +710,7 @@ _scaled_dot_product_flash_attention_nestedtensor_cuda(
       max_seqlen_batch_kv,
       output_shape) = sdpa_nested_preprocessing(query, key, value);
 
-  Tensor attention, log_sumexp, debug_attn_mask;
-  int64_t philox_seed{0}, philox_offset{0};
+  Tensor attention, log_sumexp, debug_attn_mask, philox_seed, philox_offset;
   std::tie(attention, log_sumexp, philox_seed, philox_offset, debug_attn_mask) =
```
Comment (drisspg) [inline aten/src/ATen/native/nested/cuda/NestedTensorTransformerFunctions.cpp:714]: nit: another great structured bindings  location

File: aten/src/ATen/native/transformers/cuda/attention.cu [hunk 4]
```diff
@@ -826,7 +825,7 @@ std::tuple<Tensor, Tensor, int64_t, int64_t, Tensor> _flash_attention_forward(
   at::Tensor output = at::empty_like(query);
 
   Tensor logsumexp, debug_attn_mask;
```
Comment (drisspg) [inline aten/src/ATen/native/transformers/cuda/attention.cu:827]: nit: could be same line

File: aten/src/ATen/native/transformers/cuda/flash_attn/fmha_api.cpp [hunk 5]
```diff
@@ -34,6 +34,8 @@
 #include <c10/cuda/CUDAGuard.h>
 #include <ATen/NativeFunctions.h>
 #include <ATen/cuda/CUDAGraphsUtils.cuh>
+#include <ATen/ops/scalar_tensor.h>
+#include <iostream>
```
Comment (drisspg) [inline aten/src/ATen/native/transformers/cuda/flash_attn/fmha_api.cpp:38]: 🐛

File: aten/src/ATen/native/transformers/cuda/flash_attn/fmha_api.cpp [hunk 6]
```diff
@@ -317,21 +319,31 @@ mha_fwd(const at::Tensor &q,         // total_q x num_heads x head_size, total_q
...
-    uint64_t seed{0}, offset{0};
-    if( is_dropout ) {
-        TORCH_CHECK(at::cuda::currentStreamCaptureStatus() == at::cuda::CaptureStatus::None,
-        "scaled_dot_product_flash_attention does not support dropout with cuda graph capture mode enabled");
+    at::Tensor seed_t, offset_t;
+    if (is_dropout) {
+
         // See Note [Acquire lock when using random generators]
         std::lock_guard<std::mutex> lock(gen->mutex_);
         // generator_state = at::Tensor::wrap_tensor_impl(gen -> get_state());
```
Comment (drisspg) [inline aten/src/ATen/native/transformers/cuda/flash_attn/fmha_api.cpp:327]: nit: this comment can go and was added by me when attempting to it the first go around

File: aten/src/ATen/native/transformers/cuda/flash_attn/fmha_api.cpp [hunk 7]
```diff
@@ -317,21 +319,31 @@ mha_fwd(const at::Tensor &q,         // total_q x num_heads x head_size, total_q
...
         at::PhiloxCudaState philox_state = gen->philox_cuda_state(counter_offset);
-        std::tie(seed, offset) = at::cuda::philox::unpack(philox_state);
+        if (at::cuda::currentStreamCaptureStatus() == at::cuda::CaptureStatus::None) {
+          auto [seed, offset] = at::cuda::philox::unpack(philox_state);
+          seed_t = at::scalar_tensor(at::Scalar(static_cast<int64_t>(seed)), at::dtype(at::kLong));
+          offset_t = at::scalar_tensor(at::Scalar(static_cast<int64_t>(offset)), at::dtype(at::kLong));
+        } else {
+          seed_t = at::empty({}, at::dtype(at::kLong).device(at::kCUDA));
```
Comment (drisspg) [inline aten/src/ATen/native/transformers/cuda/flash_attn/fmha_api.cpp:334]: should these be size 1?

File: aten/src/ATen/native/transformers/cuda/flash_attn/fmha_api.cpp [hunk 8]
```diff
@@ -317,21 +319,31 @@ mha_fwd(const at::Tensor &q,         // total_q x num_heads x head_size, total_q
...
+    at::Tensor seed_t, offset_t;
+    if (is_dropout) {
+
         // See Note [Acquire lock when using random generators]
         std::lock_guard<std::mutex> lock(gen->mutex_);
         // generator_state = at::Tensor::wrap_tensor_impl(gen -> get_state());
         at::PhiloxCudaState philox_state = gen->philox_cuda_state(counter_offset);
-        std::tie(seed, offset) = at::cuda::philox::unpack(philox_state);
+        if (at::cuda::currentStreamCaptureStatus() == at::cuda::CaptureStatus::None) {
+          auto [seed, offset] = at::cuda::philox::unpack(philox_state);
```
Comment (drisspg) [inline aten/src/ATen/native/transformers/cuda/flash_attn/fmha_api.cpp:330]: 1 more dumb Q I have convinced myself I understand what is going on except for now I am confused on how we can call: `https://github.com/pytorch/pytorch/blob/999e17d80a8107e88c92a1019e3d7aff1d740e8c/aten/src/ATen/cuda/detail/UnpackRaw.cuh#L19` from this location seeing as this is host side code?

is there someother overload I am missing?

File: aten/src/ATen/native/native_functions.yaml [hunk 9]
```diff
@@ -14030,12 +14030,13 @@
 - func: _scaled_dot_product_attention_math(Tensor query, Tensor key, Tensor value, Tensor? attn_mask=None, float dropout_p=0.0, bool is_causal=False, Tensor? dropout_mask=None, *, float? scale=None) -> (Tensor, Tensor)
   variants: function
 
-- func: _scaled_dot_product_flash_attention(Tensor query, Tensor key, Tensor value, float dropout_p=0.0, bool is_causal=False, bool return_debug_mask=False, *, float? scale=None) -> (Tensor ouput, Tensor logsumexp, Tensor cum_seq_q, Tensor cum_seq_k, int max_q, int max_k, int philox_seed, int philox_offset, Tensor debug_attn_mask)
+- func: _scaled_dot_product_flash_attention(Tensor query, Tensor key, Tensor value, float dropout_p=0.0, bool is_causal=False, bool return_debug_mask=False, *, float? scale=None) -> (Tensor ouput, Tensor logsumexp, Tensor cum_seq_q, Tensor cum_seq_k, int max_q, int max_k, Tensor philox_seed, Tensor philox_offset, Tensor debug_attn_mask)
```
[the remark this answers is not in the document]
Comment (ngimel) [reply aten/src/ATen/native/native_functions.yaml:14036]: Right, but we cannot return `generator` object in native functions, so we have to make do with tensors.

File: aten/src/ATen/native/transformers/cuda/flash_attn/fmha_api.cpp (hunk 8 above)
Comment (ngimel) [reply aten/src/ATen/native/transformers/cuda/flash_attn/fmha_api.cpp:330]: That's an excellent question that puzzled both me and @mcarilli, but apparently somehow it works.

**Review by ngimel — COMMENTED** [review]
I've checked that calls to sdpa are serialized to `scaled_dot_product_attention`, not `_` APIs that I'm changing, so we can try landing as is, if something breaks we can go with plan B of leaving old functions as is, and adding new ones.

**Review by drisspg — APPROVED** [review]
🚀
